dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/dataelement/DataElementMergeService.java

```diff
@@ -0,0 +1,226 @@
+  private ImmutableList<DataElementAuditMergeHandler> auditMergeHandlers;
+
+  @Override
+  public MergeRequest validate(@Nonnull MergeParams params, @Nonnull MergeReport mergeReport) {
+    log.info("Validating DataElement merge request");
+    if (params.getDataMergeStrategy() == null) {
+      mergeReport.addErrorMessage(new ErrorMessage(ErrorCode.E1556));
+      return null;
+    }
+
+    // sources
+    Set<UID> sources = new HashSet<>();
+    validator.verifySources(params.getSources(), sources, mergeReport, DataElement.class);
+
+    // target
+    validator.checkIsTargetInSources(sources, params.getTarget(), mergeReport, DataElement.class);
+    MergeRequest request = validator.verifyTarget(mergeReport, sources, params, DataElement.class);
+
+    if (mergeReport.hasErrorMessages()) return request;
+
+    // get DEs for further type-specific validation
+    DataElement deTarget = dataElementService.getDataElement(request.getTarget().getValue());
+    List<DataElement> deSources =
+        dataElementService.getDataElementsByUid(
+            request.getSources().stream().map(UID::getValue).toList());
+
+    MergeReport report =
+        dataElementMergeValidator.validateValueType(deTarget, deSources, mergeReport);
+
+    if (report.hasErrorMessages()) return request;
+
+    dataElementMergeValidator.validateDomainType(deTarget, deSources, mergeReport);
+    return request;
+  }
+
+  @Override
+  @Transactional
+  public MergeReport merge(@Nonnull MergeRequest request, @Nonnull MergeReport mergeReport) {
+    log.info("Performing DataElement merge");
+
+    List<DataElement> sources =
+        dataElementService.getDataElementsByUid(UID.toValueList(request.getSources()));
+    DataElement target = dataElementService.getDataElement(request.getTarget().getValue());
+
+    // merge metadata
+    log.info("Handling DataElement reference associations and merges");
+    dataMergeHandlers.forEach(h -> h.merge(sources, target, request));
+    metadataMergeHandlers.forEach(h -> h.merge(sources, target));
+    commonMergeHandlers.forEach(h -> h.merge(sources, target));
+    auditMergeHandlers.forEach(h -> h.merge(sources, request));
+
+    // a flush is required here to bring Hibernate into a consistent state, due to some of the
+    // above merge operations involving required native queries (queries involving JSONB). This
+    // eliminates possible stale state issues.
+    entityManager.flush();
+
+    // handle deletes
+    if (request.isDeleteSources()) handleDeleteSources(sources, mergeReport);
+
+    return mergeReport;
+  }
+
+  private void handleDeleteSources(List<DataElement> sources, MergeReport mergeReport) {
+    log.info("Deleting source DataElements");
+    for (DataElement source : sources) {
+      mergeReport.addDeletedSource(source.getUid());
+      dataElementService.deleteDataElement(source);
+    }
+  }
+
+  @PostConstruct
+  private void initMergeHandlers() {
+    metadataMergeHandlers =
```

Comment:
- I wonder if the requirement has a use case for skipping some of the merge steps in this list. Maybe for next version.

Reply:
possibly, although it seems to be an all or nothing feature, especially if the source `DataElement`s are being deleted (no refs can be left behind in the system)